Adding U-Flow method

mtailanian opened the pull request:
## Description

Adding UFlow model, solving https://github.com/openvinotoolkit/anomalib/discussions/1367#discussion-5673646

## Changes

<details>
<summary>Describe the changes you made</summary>

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Refactor (non-breaking change which refactors the code base)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update

</details>

## Checklist

<details>
<summary>Ensure that you followed the following</summary>

- [ ] I have added a summary of my changes to the [CHANGELOG](https://github.com/openvinotoolkit/anomalib/blob/main/CHANGELOG.md) (not for minor changes, docs and tests).
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas (If applicable)
- [x] I have made corresponding changes to the documentation (If applicable)
- [x] I have added tests that prove my fix is effective or that my feature works (If applicable)

</details>

samet-akcay commented on src/anomalib/models/uflow/anomaly_map.py:
```diff
@@ -0,0 +1,160 @@
+from typing import List
+
+import numpy as np
+import scipy.integrate as integrate
```
as per the codacy comment,
```suggestion
from scipy import integrate
```

samet-akcay commented on src/anomalib/models/uflow/anomaly_map.py:
```diff
@@ -0,0 +1,160 @@
+        return anomaly_mask
+
+    @staticmethod
+    def binomial_test(z: torch.Tensor, win, probability_thr: float, high_precision: bool = False):
```
```suggestion
    def binomial_test(z: torch.Tensor, win, probability_thr: float, high_precision: bool = False) -> torch.Tensor:
```

samet-akcay commented on src/anomalib/models/uflow/feature_extraction.py:
```diff
@@ -0,0 +1,141 @@
+import pytorch_lightning as pl
+import timm
+import torch
+import torch.nn as nn
```
same as above, 
```suggestion
from torch import nn
```

samet-akcay commented on src/anomalib/models/uflow/feature_extraction.py:
```diff
@@ -0,0 +1,141 @@
+        raise NotImplementedError
+
+
+class ResNetFeatureExtractor(FeatureExtractorInterface):
```
Would it be possible to use the feature extractor that is already in [here](https://github.com/openvinotoolkit/anomalib/tree/main/src/anomalib/models/components/feature_extractors)?

mtailanian replied:
Let me take a look; I think I can integrate it better...

samet-akcay commented on src/anomalib/models/uflow/README.md:
```diff
@@ -0,0 +1,109 @@
+## Localization results
+### Pixel AUROC over MVTec-AD Dataset
+
+![Pixel-AUROC results](https://raw.githubusercontent.com/openvinotoolkit/anomalib/main/docs/source/images/uflow/results/pixel-auroc.png "Pixel-AUROC results")
```
is this part of the pr, I cannot see that under `docs/source/images/uflow`

mtailanian replied:
Sorry, I missed that... I put some images inside a `results` folder that was ignored...

samet-akcay commented on src/anomalib/models/uflow/README.md:
```diff
@@ -0,0 +1,109 @@
+
+### Pixel AUPRO over MVTec-AD Dataset
+
+![Pixel-AUPRO results](https://raw.githubusercontent.com/openvinotoolkit/anomalib/main/docs/source/images/uflow/results/pixel-aupro.png "Pixel-AUPRO results")
```
Same as above, I can only see `diagram.png` and `teaser.jpg`

mtailanian replied:
Same!

samet-akcay reviewed (changes requested):
Thanks @mtailanian for your contribution! This is great! I only have some minor comments

mtailanian commented on src/anomalib/models/uflow/lightning_model.py:
```diff
@@ -0,0 +1,124 @@
+        )
+        self.loss = UFlowLoss()
+
+    def step(self, batch):
```
Sure! Will change it

mtailanian commented on src/anomalib/models/uflow/README.md:
```diff
@@ -0,0 +1,109 @@
+
+## Benchmarking
+
+Note that the proposed method uses the MCait Feature Extractor, which has an input size of 448x448. In the benchmarking, a size of 256x256 is used for all methods, and therefore the results may differ from those reported. In order to exactly reproduce all results, the reader can refer to the original code (see [here](https://www.github.com/mtailanian/uflow), where the configs used and even the tr ...
```
I was running the tests using a `original_size -> 256 -> 448` strategy, and the performance dropped a bit... 
I think we can match the paper's numbers with this code, but I will double-check it!

mtailanian commented:
Hi @samet-akcay ! Thanks for all your comments! 
This new version I just pushed should be addressing all of them! 

Let me know if you have further comments! 

Thanks! :)

djdameln commented on src/anomalib/models/uflow/torch_model.py:
```diff
@@ -0,0 +1,134 @@
+
+        return ff.GraphINN(input_nodes + nodes + output_nodes[::-1])
+
+    def get_flow_stage(self, in_node: ff.Node, flow_steps: int, condition_node: ff.Node = None) -> list[ff.Node]:
```
would it be more accurate to call this `build_flow_stage`?

mtailanian replied:
Yes, agree!

ashwinvaidya17 reviewed (commented):
Thanks. Minor comments

mtailanian commented on src/anomalib/models/uflow/lightning_model.py:
```diff
@@ -0,0 +1,120 @@
+                drop_last = 1 * self.trainer.train_dataloader.loaders.drop_last
+                iterations_per_epoch = number_of_training_examples // batch_size + 1 - drop_last
+                total_iterations = iterations_per_epoch * (self.trainer.max_epochs - 1)
+            except:
```
Honestly, I'm not being able to reproduce this error... At some point (as far as I remember), the error was generated because, at this point, the dataloaders are not yet initialized. It seems like they are initialized in the .fit loop (according to [this](https://github.com/Lightning-AI/lightning/issues/10430#issue-1048444548)).

mtailanian commented on src/anomalib/models/uflow/lightning_model.py:
```diff
@@ -0,0 +1,120 @@
+__all__ = ["Uflow", "UflowLightning"]
+
+
+class Uflow(AnomalyModule):
```
Sure!

mtailanian commented:
Hello guys! Thanks for your comments. The last commits should address everything!
Thanks!

samet-akcay reviewed (approved):
thanks @mtailanian, lgtm!

samet-akcay commented:
@mtailanian, code quality checks are failing. Will you be able to run `pre-commit run --all-files` to address these issues. We could then merge it?

mtailanian commented:
Done! 
Thanks!

samet-akcay commented:
Looks like one of the tests fail when running the inference. Can you see the logs? If not I can paste the output here?

mtailanian commented:
Hi @samet-akcay , how are you?
I was having a problem exporting the model to Torch because the network was not "pickable", because:
* It was using lambda functions
* And a function inside a function

The changes I made to fix it are the following:
* I swapped Freia's AllInOneBlock for Anomalib's version: fixed lambda stuff
* Changed the function-in-function stuff to a Class. 

I think everything should be fine now... The `torch_inference` is running fine... 
Please let me know in any case...

samet-akcay commented:
@mtailanian, thanks. It works now! Thanks for your effort.

Yeah, torch inference does not like lambda functions. That is the reason why modified some of the freia modules. Maybe we could create a pr on their side. -this is a todo task for us :)

I'm happy to merge your PR! Thanks again for your contribution!

mtailanian commented:
Yay! :) 
Awesome! Thank you very much @samet-akcay , and all!
